File: Sources/PetStateMachine.swift
import Foundation
import SwiftUI

enum PetState: Equatable {
    case idle
    case looking(Direction)
    case dragging
    case typing(heat: Double)
    case petting
    case stretching
    case stretchReminder
    case waterReminder
}

enum StretchInterval: Equatable {
    case off
    case custom(Int)
    
    var rawValue: Int {
        switch self {
        case .off: return 0
        case .custom(let secs): return secs
        }
    }
}

enum WaterInterval: Equatable {
    case off
    case custom(Int)
    
    var rawValue: Int {
        switch self {
        case .off: return 0
        case .custom(let secs): return secs
        }
    }
}

enum PawSide: Equatable {
    case left
    case right
}

enum Direction {
    case up, down, left, right, upLeft, upRight, downLeft, downRight, center
}

class PetStateMachine: ObservableObject {
    @Published var currentState: PetState = .idle {
        didSet {
            let isStretching = (currentState == .stretchReminder)
            let wasStretching = (oldValue == .stretchReminder)
            if isStretching != wasStretching {
                withAnimation(.easeInOut(duration: 0.3)) {
                    stretchColorAmount = isStretching ? 1.0 : 0.0
                }
            }
            
            let isWatering = (currentState == .waterReminder)
            let wasWatering = (oldValue == .waterReminder)
            if isWatering != wasWatering {
                withAnimation(.easeInOut(duration: 0.3)) {
                    waterColorAmount = isWatering ? 1.0 : 0.0
                }
            }
        }
    }
    @Published var isHiding: Bool = false
    @Published var isBlipping: Bool = false
    @Published var typingHeat: Double = 0.0 // 0.0 to 1.0
    
    @Published var stretchColorAmount: Double = 0.0
    @Published var waterColorAmount: Double = 0.0
    @Published var wagTick: Bool = false
    @Published var lastTypedPaw: PawSide = .left
    
    @Published var basePurrMessage: String = "" {
        didSet {
            UserDefaults.standard.set(basePurrMessage, forKey: "userPurrMessage")
            if currentState != .stretchReminder && currentState != .waterReminder {
                purrMessage = basePurrMessage
            }
        }
    }
    @Published var baseShowPurrMessage: Bool = false {
        didSet {
            UserDefaults.standard.set(baseShowPurrMessage, forKey: "userShowPurrMessage")
            if currentState != .stretchReminder && currentState != .waterReminder {
                showPurrMessage = baseShowPurrMessage
            }
        }
    }
    
    @Published var purrMessage: String = ""
    @Published var showPurrMessage: Bool = false
    
    @Published var isPomodoroActive: Bool = false
    @Published var pomodoroTimeRemaining: Int = 0
    private var pomodoroDuration: Int = 25 * 60
    
    @Published var stretchInterval: StretchInterval = .off {
        didSet {
            UserDefaults.standard.set(stretchInterval.rawValue, forKey: "stretchInterval")
            resetStretchTimer()
        }
    }
    @Published var timeUntilStretch: Int = 0

    @Published var waterInterval: WaterInterval = .off {
        didSet {
            UserDefaults.standard.set(waterInterval.rawValue, forKey: "waterInterval")
            resetWaterTimer()
        }
    }
    @Published var timeUntilWater: Int = 0

    private var lastDirectionChange: Date = Date()
    private let debounceInterval: TimeInterval = 0.05
    
    private var keystrokes: [Date] = []
    private var stateTimer: Timer?
    private var pettingTimer: Timer?
    private var lastInteractionTime: Date = Date()
    private var pettingAccumulator: CGFloat = 0
    private var pettingStartTime: Date?
    
    private var isCurrentlyTyping: Bool = false
    private var typingTimer: Timer?
    
    private var wagTimer: Timer?
    
    init() {
        let defaults = UserDefaults.standard
        
        if let savedPurr = defaults.string(forKey: "userPurrMessage") {
            self.basePurrMessage = savedPurr
            self.baseShowPurrMessage = defaults.bool(forKey: "userShowPurrMessage")
            self.purrMessage = self.basePurrMessage
            self.showPurrMessage = self.baseShowPurrMessage
        }
        
        if defaults.object(forKey: "waterInterval") == nil {
            self.waterInterval = .custom(60 * 60)
        } else {
            let val = defaults.integer(forKey: "waterInterval")
            self.waterInterval = val == 0 ? .off : .custom(val)
        }
        
        if defaults.object(forKey: "stretchInterval") == nil {
            self.stretchInterval = .custom(45 * 60)
        } else {
            let val = defaults.integer(forKey: "stretchInterval")
            self.stretchInterval = val == 0 ? .off : .custom(val)
        }

        self.timeUntilWater = self.waterInterval.rawValue
        self.timeUntilStretch = self.stretchInterval.rawValue

        stateTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            guard let self = self else { return }
            self.updateTypingHeat()
            self.updatePomodoro()
            self.updateStretchReminder()
            self.updateWaterReminder()
            
            // Random interactions roughly every 5 seconds
            if Int.random(in: 1...5) == 1 {
                self.updateIdleState()
            }
        }
        
        wagTimer = Timer.scheduledTimer(withTimeInterval: 0.15, repeats: true) { [weak self] _ in
            self?.wagTick.toggle()
        }
    }
    
    func registerInteraction() {
        lastInteractionTime = Date()
    }
    
    func registerPetting(delta: CGFloat) {
        registerInteraction()
        
        pettingAccumulator += delta
        
        pettingTimer?.invalidate()
        pettingTimer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: false) { [weak self] _ in
            DispatchQueue.main.async {
                self?.stopPetting()
            }
        }
        
        if pettingStartTime == nil {
            pettingStartTime = Date()
        }
        
        if let startTime = pettingStartTime, Date().timeIntervalSince(startTime) >= 1.2 {
            if pettingAccumulator > 30 {
                DispatchQueue.main.async {
                    if self.currentState != .dragging && self.currentState != .petting {
                        if case .typing = self.currentState { return }
                        self.currentState = .petting
                    }
                }
            }
        }
    }
    
    func stopPetting() {
        pettingAccumulator = 0
        pettingStartTime = nil
        if case .petting = currentState {
            currentState = .idle
        }
    }
    
    func wakeUpIfNeeded() {
        if isHiding {
            DispatchQueue.main.async {
                self.isHiding = false
                if self.currentState != .stretchReminder && self.currentState != .waterReminder {
                    self.currentState = .idle
                }
            }
        }
    }
    
    func startPomodoro() {
        DispatchQueue.main.async {
            self.isPomodoroActive = true
            self.pomodoroTimeRemaining = self.pomodoroDuration
        }
    }
    
    func stopPomodoro() {
        DispatchQueue.main.async {
            self.isPomodoroActive = false
        }
    }
    
    private func updatePomodoro() {
        guard isPomodoroActive else { return }
        if pomodoroTimeRemaining > 0 {
            pomodoroTimeRemaining -= 1
        } else if pomodoroTimeRemaining == 0 {
            isPomodoroActive = false
            purrMessage = "Pomodoro Done! 🐾"
            showPurrMessage = true
        }
    }
    
    private func updateStretchReminder() {
        if stretchInterval == .off { return }
        
        if timeUntilStretch > 0 {
            timeUntilStretch -= 1
        } else {
            if currentState != .stretchReminder {
                self.wakeUpIfNeeded()
                currentState = .stretchReminder
                purrMessage = "Time to stretch! 🐾"
                showPurrMessage = true
            }
        }
    }
    
    func resetStretchTimer() {
        let wasStretchReminder = (currentState == .stretchReminder)
        DispatchQueue.main.async {
            self.timeUntilStretch = self.stretchInterval.rawValue
            if wasStretchReminder {
                self.currentState = .idle
                self.purrMessage = self.basePurrMessage
                self.showPurrMessage = self.baseShowPurrMessage
            }
        }
    }
    
    func acknowledgeStretch() {
        if currentState == .stretchReminder {
            resetStretchTimer()
        }
    }

    private func updateWaterReminder() {
        if waterInterval == .off { return }
        
        if timeUntilWater > 0 {
            timeUntilWater -= 1
        } else {
            if currentState != .waterReminder {
                self.wakeUpIfNeeded()
                currentState = .waterReminder
                purrMessage = "Drink Water!"
                showPurrMessage = true
            }
        }
    }
    
    func resetWaterTimer() {
        let wasWaterReminder = (currentState == .waterReminder)
        DispatchQueue.main.async {
            self.timeUntilWater = self.waterInterval.rawValue
            if wasWaterReminder {
                self.currentState = .idle
                self.purrMessage = self.basePurrMessage
                self.showPurrMessage = self.baseShowPurrMessage
            }
        }
    }
    
    func acknowledgeWater() {
        if currentState == .waterReminder {
            resetWaterTimer()
        }
    }
    
    private func updateIdleState() {
        // Only trigger random states if we are currently idle and not blipping or hiding
        guard case .idle = currentState, !isHiding, !isBlipping else { return }
        
        // Very low chance to do a random stretch
        if Double.random(in: 0...1) < 0.05 {
            currentState = .stretching
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
                if case .stretching = self.currentState {
                    self.currentState = .idle
                }
            }
        }
    }
    
    func updateDirection(to newDirection: Direction) {
        registerInteraction()
        
        if case .dragging = currentState { return }
        if case .typing = currentState { return }
        if case .petting = currentState { return }
        if case .stretching = currentState { return }
        if case .stretchReminder = currentState { return }
        if case .waterReminder = currentState { return }
        
        let now = Date()
        if now.timeIntervalSince(lastDirectionChange) > debounceInterval {
            DispatchQueue.main.async {
                self.currentState = .looking(newDirection)
                self.lastDirectionChange = now
            }
        }
    }
    
    func startDragging() {
        registerInteraction()
        DispatchQueue.main.async {
            self.currentState = .dragging
            self.isHiding = false
            self.typingHeat = 0.0
        }
    }
    
    func stopDragging() {
        DispatchQueue.main.async {
            if case .dragging = self.currentState {
                self.currentState = .idle
            }
        }
    }
    
    func registerKeystroke() {
        registerInteraction()
        wakeUpIfNeeded()
        let now = Date()
        keystrokes.append(now)
        
        lastTypedPaw = (lastTypedPaw == .left) ? .right : .left
        
        DispatchQueue.main.async {
            self.isCurrentlyTyping = true
            self.typingTimer?.invalidate()
            self.typingTimer = Timer.scheduledTimer(withTimeInterval: 0.5, repeats: false) { [weak self] _ in
                DispatchQueue.main.async {
                    self?.isCurrentlyTyping = false
                    self?.updateTypingState()
                }
            }
            self.updateTypingHeat()
        }
    }
    
    private func updateTypingHeat() {
        let now = Date()
        keystrokes = keystrokes.filter { now.timeIntervalSince($0) <= 3.0 }
        
        DispatchQueue.main.async {
            self.recalculateHeat()
        }
        
        // Check for idle hiding
        if now.timeIntervalSince(lastInteractionTime) > 15.0 {
            DispatchQueue.main.async {
                if !self.isHiding && self.currentState != .dragging && self.currentState != .stretchReminder && self.currentState != .waterReminder {
                    self.isHiding = true
                }
            }
        }
    }
    
    private func recalculateHeat() {
        let kps = Double(keystrokes.count) / 3.0
        
        let minKps = 1.25 // 15 WPM
        let maxKps = 5.83 // 70 WPM
        
        let newHeat: Double
        if kps < minKps {
            newHeat = 0.0
        } else if kps >= maxKps {
            newHeat = 1.0
        } else {
            newHeat = (kps - minKps) / (maxKps - minKps)
        }
        
        withAnimation(.linear(duration: 0.2)) {
            typingHeat = newHeat
        }
        
        updateTypingState()
    }
    
    private func updateTypingState() {
        if isCurrentlyTyping {
            if currentState != .stretchReminder && currentState != .waterReminder && currentState != .dragging && currentState != .petting {
                currentState = .typing(heat: typingHeat)
            }
        } else {
            if case .typing = currentState {
                currentState = .idle
            }
        }
    }
}

File: windows/pet_state_machine.py
import time
import random
import threading
from typing import List

class Direction:
    UP = "up"
    DOWN = "down"
    LEFT = "left"
    RIGHT = "right"
    UP_LEFT = "upLeft"
    UP_RIGHT = "upRight"
    DOWN_LEFT = "downLeft"
    DOWN_RIGHT = "downRight"
    CENTER = "center"

class PawSide:
    LEFT = "left"
    RIGHT = "right"

class PetState:
    IDLE = "idle"
    LOOKING = "looking"
    DRAGGING = "dragging"
    TYPING = "typing"
    PETTING = "petting"
    STRETCHING = "stretching"
    STRETCH_REMINDER = "stretchReminder"
    WATER_REMINDER = "waterReminder"

def get_state_priority(state: str) -> int:
    priorities = {
        PetState.DRAGGING: 100,
        PetState.STRETCH_REMINDER: 90,
        PetState.WATER_REMINDER: 80,
        PetState.PETTING: 70,
        PetState.TYPING: 60,
        PetState.LOOKING: 50,
        PetState.STRETCHING: 40,
        PetState.IDLE: 0
    }
    return priorities.get(state, 0)

class PetStateMachine:
    def __init__(self, update_callback=None):
        self.update_callback = update_callback
        self.current_state = PetState.IDLE
        self.looking_direction = Direction.CENTER
        self.is_hiding = False
        self.is_blipping = False
        self.typing_heat = 0.0
        self.stretch_color_amount = 0.0
        self.water_color_amount = 0.0
        self.wag_tick = False
        self.last_typed_paw = PawSide.LEFT

        self.base_purr_message = ""
        self.base_show_purr_message = False
        self.purr_message = ""
        self.show_purr_message = False

        self.is_pomodoro_active = False
        self.pomodoro_time_remaining = 0
        self.pomodoro_duration = 25 * 60

        self.stretch_interval = 45 * 60  # seconds, 0 = off
        self.time_until_stretch = self.stretch_interval
        self.water_interval = 60 * 60  # seconds, 0 = off
        self.time_until_water = self.water_interval

        self.last_direction_change = time.time()
        self.keystrokes: List[float] = []
        self.last_interaction_time = time.time()

        self.petting_accumulator = 0.0
        self.petting_start_time = None
        self.petting_timer = None

        self.is_currently_typing = False
        self.typing_timer = None

        self._start_timers()

    def _notify(self):
        if self.update_callback:
            self.update_callback()

    def set_state(self, new_state, force=False):
        if force or get_state_priority(new_state) >= get_state_priority(self.current_state):
            old_state = self.current_state
            self.current_state = new_state
            
            # handle stretch color amount
            is_stretching = (self.current_state == PetState.STRETCH_REMINDER)
            self.stretch_color_amount = 1.0 if is_stretching else 0.0
            
            # handle water color amount
            is_watering = (self.current_state == PetState.WATER_REMINDER)
            self.water_color_amount = 1.0 if is_watering else 0.0
            
            self._notify()
            return True
        return False

    def register_interaction(self):
        self.last_interaction_time = time.time()

    def register_petting(self, delta: float):
        self.register_interaction()
        self.petting_accumulator += delta
        
        if self.petting_timer:
            self.petting_timer.cancel()
            
        def stop_pet():
            self.stop_petting()
        self.petting_timer = threading.Timer(0.5, stop_pet)
        self.petting_timer.start()

        if self.petting_start_time is None:
            self.petting_start_time = time.time()

        if time.time() - self.petting_start_time >= 1.2:
            if self.petting_accumulator > 30:
                self.set_state(PetState.PETTING)

    def stop_petting(self):
        self.petting_accumulator = 0
        self.petting_start_time = None
        if self.current_state == PetState.PETTING:
            self.set_state(PetState.IDLE, force=True)

    def wake_up_if_needed(self):
        if self.is_hiding:
            self.is_hiding = False
            if self.current_state not in [PetState.STRETCH_REMINDER, PetState.WATER_REMINDER]:
                self.set_state(PetState.IDLE)
            self._notify()

    def start_pomodoro(self):
        self.is_pomodoro_active = True
        self.pomodoro_time_remaining = self.pomodoro_duration
        self._notify()

    def stop_pomodoro(self):
        self.is_pomodoro_active = False
        self._notify()

    def acknowledge_stretch(self):
        if self.current_state == PetState.STRETCH_REMINDER:
            self.time_until_stretch = self.stretch_interval
            self.set_state(PetState.IDLE, force=True)
            self.purr_message = self.base_purr_message
            self.show_purr_message = self.base_show_purr_message
            self._notify()

    def acknowledge_water(self):
        if self.current_state == PetState.WATER_REMINDER:
            self.time_until_water = self.water_interval
            self.set_state(PetState.IDLE, force=True)
            self.purr_message = self.base_purr_message
            self.show_purr_message = self.base_show_purr_message
            self._notify()

    def acknowledge_pomodoro(self):
        if self.purr_message == "Pomodoro Done! 🐾":
            self.purr_message = self.base_purr_message
            self.show_purr_message = self.base_show_purr_message
            self._notify()

    def update_direction(self, new_direction):
        self.register_interaction()
        now = time.time()
        if now - self.last_direction_change > 0.05:
            if self.set_state(PetState.LOOKING):
                self.looking_direction = new_direction
                self.last_direction_change = now

    def start_dragging(self):
        self.register_interaction()
        self.set_state(PetState.DRAGGING, force=True)
        self.is_hiding = False
        self.typing_heat = 0.0
        self._notify()

    def stop_dragging(self):
        if self.current_state == PetState.DRAGGING:
            self.set_state(PetState.IDLE, force=True)

    def register_keystroke(self):
        self.register_interaction()
        self.wake_up_if_needed()
        self.keystrokes.append(time.time())

        self.last_typed_paw = PawSide.RIGHT if self.last_typed_paw == PawSide.LEFT else PawSide.LEFT
        self.is_currently_typing = True

        if self.typing_timer:
            self.typing_timer.cancel()
            
        def stop_typing():
            self.is_currently_typing = False
            self.update_typing_state()
        self.typing_timer = threading.Timer(0.5, stop_typing)
        self.typing_timer.start()
        
        self.update_typing_heat()

    def update_typing_heat(self):
        now = time.time()
        self.keystrokes = [t for t in self.keystrokes if now - t <= 3.0]
        
        kps = len(self.keystrokes) / 3.0
        min_kps = 1.25  # 15 WPM
        max_kps = 5.83  # 70 WPM

        if kps < min_kps:
            self.typing_heat = 0.0
        elif kps >= max_kps:
            self.typing_heat = 1.0
        else:
            self.typing_heat = (kps - min_kps) / (max_kps - min_kps)

        self.update_typing_state()

        if now - self.last_interaction_time > 15.0:
            if not self.is_hiding and self.current_state not in [PetState.DRAGGING, PetState.STRETCH_REMINDER, PetState.WATER_REMINDER]:
                self.is_hiding = True
                self._notify()

    def update_typing_state(self):
        if self.is_currently_typing:
            self.set_state(PetState.TYPING)
        else:
            if self.current_state == PetState.TYPING:
                self.set_state(PetState.IDLE, force=True)
        self._notify()

    def _tick_1s(self):
        self.update_typing_heat()
        
        if self.is_pomodoro_active:
            if self.pomodoro_time_remaining > 0:
                self.pomodoro_time_remaining -= 1
            elif self.pomodoro_time_remaining == 0:
                self.is_pomodoro_active = False
                self.purr_message = "Pomodoro Done! 🐾"
                self.show_purr_message = True
                self._notify()

        if self.stretch_interval > 0:
            if self.time_until_stretch > 0:
                self.time_until_stretch -= 1
            else:
                if self.set_state(PetState.STRETCH_REMINDER):
                    self.wake_up_if_needed()
                    self.purr_message = "Time to stretch! 🐾"
                    self.show_purr_message = True

        if self.water_interval > 0:
            if self.time_until_water > 0:
                self.time_until_water -= 1
            else:
                if self.set_state(PetState.WATER_REMINDER):
                    self.wake_up_if_needed()
                    self.purr_message = "Drink Water!"
                    self.show_purr_message = True

        if self.current_state == PetState.IDLE and not self.is_hiding and not self.is_blipping:
            if random.randint(1, 5) == 1:
                if random.random() < 0.05:
                    if self.set_state(PetState.STRETCHING):
                        def end_stretch():
                            if self.current_state == PetState.STRETCHING:
                                self.set_state(PetState.IDLE, force=True)
                        threading.Timer(1.5, end_stretch).start()

        threading.Timer(1.0, self._tick_1s).start()

    def _tick_0_15s(self):
        self.wag_tick = not self.wag_tick
        self._notify()
        threading.Timer(0.15, self._tick_0_15s).start()

    def _start_timers(self):
        threading.Timer(1.0, self._tick_1s).start()
        threading.Timer(0.15, self._tick_0_15s).start()
